src/core/middleware/theme.ts

```diff
@@ -0,0 +1,99 @@
+import { create, invalidator, diffProperties, getRegistry } from '../vdom';
+import cache from './cache';
+import injector from './injector';
+import Injector from '../Injector';
+import Set from '../../shim/Set';
+import { registerThemeInjector, Theme, Classes, ClassNames, INJECTED_THEME_KEY, THEME_KEY } from '../mixins/Themed';
+import { shallow } from '../diff';
+
+export interface ThemedProperties {
+	theme?: Theme;
+	classes?: Classes;
+}
+
+const factory = create({ invalidator, cache, diffProperties, injector, getRegistry }).properties<ThemedProperties>();
+
+export const theme = factory(
+	({ middleware: { invalidator, cache, diffProperties, injector, getRegistry }, properties }) => {
+		let themeKeys = new Set();
+		diffProperties((current: ThemedProperties, next: ThemedProperties) => {
+			let result = false;
+			if (
+				current.theme !== next.theme ||
+				(current.classes && !next.classes) ||
+				(!current.classes && next.classes)
+			) {
+				result = true;
+			} else if (current.classes && next.classes) {
+				const keys = [...themeKeys.values()];
+				for (let i = 0; i < keys.length; i++) {
+					let key = keys[i];
+					result = shallow(current.classes[key], next.classes[key], 1).changed;
```

Comment:
Diffs the classes for every key that has been registered by the widget, the classes object is nested three levels so requires a slightly deeper diff than the standard `shallow` diff. To support this `shallow` accepts a depth param to control the level of the diff.